Tweak build rules for WebAssembly

ChinYikMing opened the pull request:
The "emcc --version" command displays information about the version of emcc, which contains the "gcc" or "clang" string. As a result, the build rule should only match one compiler and stop checking anymore.  If not, the build could be failed by a mix of different compiler's CFLAGS. For example, -flto from clang and -flto=thin from emcc.

After refactoring the riscv.[ch] public APIs in commit 820cd9b, it is now possible to adjust the memory size to accommodate varying runtime requirements. The memory size must align to a 4KB page size for the WASM runtime and I set 1GB as default memory size for WASM runtime.

The LTO is not supported to build emscripten-port SDL library, so disable it when both ENABLE_LTO=1 and ENABLE_SDL=1 are set.

Related to: #75
Close #375

jserv commented on Makefile:
```diff
@@ -167,13 +190,25 @@ OBJS := \
+EMCC_CFLAGS :=
+EMCC_MACRO :=
+ifeq ("$(CC_IS_EMCC)", "1")
+EMCC_MACRO += -DMEM_SIZE=0x40000000    # 1GB
```
The name `EMCC_MACRO` is misleading. Drop it and move to `CFLAGS_emcc`.

jserv replied:
Instead of "1GB," write "1 GiB" in [binary prefix](https://en.wikipedia.org/wiki/Binary_prefix) form.

jserv commented on Makefile:
```diff
@@ -167,13 +190,25 @@ OBJS := \
+EMCC_MACRO :=
+ifeq ("$(CC_IS_EMCC)", "1")
+EMCC_MACRO += -DMEM_SIZE=0x40000000    # 1GB
+EMCC_CFLAGS += -sINITIAL_MEMORY=2GB \
```
Rename to `CFLAGS_emcc` to emphasize the use of emcc-specific C flags.

jserv commented on Makefile:
```diff
@@ -167,13 +190,25 @@ OBJS := \
+	       -sALLOW_MEMORY_GROWTH \
+	       --embed-file build \
+	       -s"EXPORTED_FUNCTIONS=$(EXPORTED_FUNCS)" \
+	       -sMALLOC=mimalloc
```
Since you are using mimalloc, you have to check the versions of emcc which bundle the corresponding packages.

jserv commented on Makefile:
```diff
@@ -148,6 +164,13 @@ $(OUT)/emulate.o: CFLAGS += -foptimize-sibling-calls -fomit-frame-pointer -fno-s
 # to the first target after .DEFAULT_GOAL is not set.
 .DEFAULT_GOAL :=
 
+EMCC_OBJS += $(BIN).js \
```
It is confusing to name them object files.

ChinYikMing replied:
> It is confusing to name them object files.

What about `EMCC_OUT`, these files are generated by `emcc`.

jserv replied:
Use `WEB_FILES` instead.

jserv commented on Makefile:
```diff
@@ -136,6 +143,15 @@ ifeq ("$(CC_IS_EMCC)", "1")
+# Build emscripten-port SDL
+ifeq ("$(CC_IS_EMCC)", "1")
+ifeq ($(call has, SDL), 1)
+EMCC_PORT_FLAGS := -sUSE_SDL=2 -sSDL2_MIXER_FORMATS=wav,mid -sUSE_SDL_MIXER=2
```
Simply name it `CFLAGS_emcc_port`.

ChinYikMing replied:
> Simply name it `CFLAGS_emcc_port`.

No more need  `CFLAGS_emcc_port` since we can simply concatenate flags to `CFLAGS_emcc`.

jserv commented on Makefile:
```diff
@@ -167,13 +190,47 @@ OBJS := \
+	       -DMEM_SIZE=0x40000000 \
+	       -w
+
+# see commit 165c1a3 of emscripten
```
Move the emcc specific rules to `mk/toolchain.mk`.

ChinYikMing committed:
Tweak build rules for WebAssembly

The "emcc --version" command displays information about the version of
emcc, which contains the "gcc" or "clang" string. As a result, the build
rule should only match one compiler and stop checking anymore.  If not,
the build could be failed by a mix of different compiler's CFLAGS. For
example, -flto from clang and -flto=thin from emcc.

After refactoring the riscv.[ch] public APIs in commit 820cd9b, it is
now possible to adjust the memory size to accommodate varying runtime
requirements. The memory size must align to a 4KB page size for the
WASM runtime and I set 1GB as default memory size for WASM runtime.

The LTO is not supported to build emscripten-port SDL library,
so disable it when both ENABLE_LTO=1 and ENABLE_SDL=1 are set.

Related to: #75
Close #375

jserv commented:
Thank @ChinYikMing for contributing!
